rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,283 @@
+  op but with statically known upper bounds of `output_dimensions` operand.
+* Bounded dynamism can also open up performance optimizations opportunities.
+
+## Recommendations
+
+### StableHLO Producers
+
+* Producers can use the bounded tensor type representation as described in
+  P1.
+* Producers are encouraged to use the unbounded dynamic operations as described
+  in P4. Result types of these ops aren't required to have a bounded type.
+* Producers can still use `get_dimension_size` and `set_dimension_size` ops
+  described in P3 for the ease of transition to StableHLO and faster adoption of
+  StableHLO.
+
+### StableHLO Consumers
+
+* Consumers should aim to support unbounded programs and can optionally make use
+  of bounds on tensors for optimizations.
+* Consumers that support unbounded programs can safely ignore the bounds
+  completely without affecting the correctness.
+* Consumers that only support bounded programs could first transform the given
+  program to a bounded one through program analysis.
+* Consumers can choose to not support `get_dimension_size` and
+  `set_dimension_size` ops until they have a motivating use-case.
+
+## Non Goals
+
+* Provide [value inference](https://github.com/openxla/xla/blob/9e05932a2ceadea080dc9494cfe9d735f94c4e68/xla/client/value_inference.h)
+  like utility for producers that want to generate the `set_dimension_size` ops.
+  Value inference depends on constant folding for StableHLO ops which is a work
+  in progress currently. There will be separate RFC for value inference subject
+  to separate approval. Note that producers that generate unbounded programs
+  don't need this
+  in StableHLO.
+* Provide a transformation that converts StableHLO programs to bounded StableHLO
+  programs. However, there is a plan to have such a conversion in MHLO, although
+  the details of this are out of scope of this RFC.
+
+## Detailed Proposal
+
+### (P1) Bounded tensor type using the encoding field in the RankedTensorType
+
+Bounds of a dynamic tensor are represented using the `TypeExtensionsAttr` in the
+`RankedTensorType` encoding field. Bounds in `TypeExtensionsAttr` is an
+`int64_t` array of size equal to rank of the tensor. Values corresponding to
+static dimensions must be `ShapedType::kDynamicSize` which is printed as `?` in
+the IR. Disallowing a static bound value for static dimensions makes the IR
+canonical and makes it possible to infer that the dimension is dynamic if the
+bound value is static.
+
+The following type represents a 2D tensor, with the size of the 0th dimension
+being up to 3 and the size of the 1th dimension being exactly 5
+
+```mlir
+tensor<?x5xf32, #stablehlo.type_extensions<bounds = [3, ?]>>
+```
+
+Type compatibility in StableHLO also checks for compatibility of the bounds.
+The example type above is compatible with `tensor<2x5xf32>` but not with
```

Comment:
"For example, the example type above...".

Reply:
Done.